cmd/remote-configuration/agent/agent.go

```diff
@@ -0,0 +1,283 @@
+		InstanceUid:    agent.instanceId.String(),
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc:              agent.onConnect,
+			OnConnectFailedFunc:        agent.onConnectFailed,
+			OnErrorFunc:                agent.onError,
+			SaveRemoteConfigStatusFunc: agent.saveRemoteConfigStatus,
+			GetEffectiveConfigFunc:     agent.getEffectiveConfig,
+			OnMessageFunc:              agent.onMessage,
+		},
+		RemoteConfigStatus:    agent.remoteConfigStatus,
+		PackagesStateProvider: nil,
+		Capabilities:          agent.config.GetCapabilities(),
+	}
+	err := agent.opampClient.SetAgentDescription(agent.agentDescription)
+	if err != nil {
+		return err
+	}
+	err = agent.opampClient.SetHealth(agent.getHealth())
+	if err != nil {
+		return err
+	}
+
+	agent.logger.Debugf("Starting OpAMP client...")
+
+	err = agent.opampClient.Start(context.Background(), settings)
+	if err != nil {
+		return err
+	}
+
+	agent.logger.Debugf("OpAMP Client started.")
+
+	return nil
+}
+
+func (agent *Agent) updateAgentIdentity(instanceId ulid.ULID) {
+	agent.logger.Debugf("Agent identify is being changed from id=%v to id=%v",
+		agent.instanceId.String(),
+		instanceId.String())
+	agent.instanceId = instanceId
+
+	// TODO: reinit or update meter (possibly using a single function to update all own connection settings
+}
+
+func (agent *Agent) getNameAndNamespace(key string) (string, string, error) {
+	s := strings.Split(key, "/")
+	// We expect map keys to be of the form name/namespace
+	if len(s) != 2 {
+		return "", "", errors.New("invalid key")
+	}
+	return s[0], s[1], nil
+}
+
+func (agent *Agent) makeKeyFromNameNamespace(name string, namespace string) string {
+	return fmt.Sprintf("%s/%s", name, namespace)
+}
+
+func (agent *Agent) getEffectiveConfig(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+	instances, err := agent.applier.ListInstances()
+	if err != nil {
+		agent.logger.Errorf("couldn't list instances", err)
+		return nil, err
+	}
+	instanceMap := map[string]*protobufs.AgentConfigFile{}
+	for _, instance := range instances {
+		marshaled, err := yaml.Marshal(instance)
+		if err != nil {
+			agent.logger.Errorf("couldn't marshal collector configuration", err)
+			return nil, err
+		}
+		mapKey := agent.makeKeyFromNameNamespace(instance.GetName(), instance.GetNamespace())
+		instanceMap[mapKey] = &protobufs.AgentConfigFile{
+			Body:        marshaled,
+			ContentType: "yaml",
+		}
+	}
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: instanceMap,
+		},
+	}, nil
+}
+
+func (agent *Agent) initMeter(settings *protobufs.TelemetryConnectionSettings) {
+	reporter, err := metrics.NewMetricReporter(agent.logger, settings, agent.config.GetAgentType(), agent.config.GetAgentVersion(), agent.instanceId)
+	if err != nil {
+		agent.logger.Errorf("Cannot collect metrics: %v", err)
+		return
+	}
+
+	if agent.metricReporter != nil {
+		agent.metricReporter.Shutdown()
+	}
+	agent.metricReporter = reporter
```

Comment:
Can `initMeter()` be called concurrently? This feels like something that could be messy with concurrency and need a lock.